### PR TITLE
Con 1062 soft delete certificate

### PR DESCRIPTION
**Pull request by VasanthaKasirajan3008**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/DeleteCertificateRequest.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using MediatR;
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/DeleteCertificateRequest.cs:1]: Assuming that the Admin branch would be duplicating this code into the Admin service repo where necessary and Shoma knows this?

Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/DeleteCertificateRequest.cs:1]: I have updated Shoma on this.

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Certificates/DeleteCertificateHandlerTests/When_called_to_delete_certificate.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        [Test]
+        public async Task Then_soft_delte_certificate()
+        {
+            //Act
+            var result = await _deleteCertificateHandler.Handle(_deleteCertificateRequest, CancellationToken.None);
+
+            //Assert
+            Assert.IsNotNull(result);
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Certificates/DeleteCertificateHandlerTests/When_called_to_delete_certificate.cs:45]: done

File: src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs:1]: as we discussed, i will ask Shoma to do , if required i can help her on this

File: src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+            //Assert
+            _exception.Should().BeNull();            
+        }
+
+        [Test]
+        public async Task Then_Delete_With_ReasonForChange()
+        {
+            //Act           
+            await _certificateRepository.Delete(1111111111, 93, "UserName", CertificateActions.Delete, reasonForChange: _reasonForChange);
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs:52]: done. Thanks for your help on this.

File: src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+           
+            //Assert
+            var result = _certificateRepository.GetCertificateLogsFor(_certificateId);
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs:55]: done. Thanks for your help on this.

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs [hunk 6]
```diff
@@ -327,14 +327,26 @@ public async Task Delete(long uln, int standardCode, string username, string act
...
 
+            if (incidentNumber != null)
+            {
+                UpdateIncidentNumber(incidentNumber, cert);
+            }
+
             if (updateLog)
             {
-                await UpdateCertificateLog(cert, action, username);
+                await UpdateCertificateLog(cert, action, username, reasonForChange);
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Data/CertificateRepository.cs:337]: as we discussed, no change required

File: src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            _mockDbContext = CreateMockDbContext(mockCertificate, mockCertificateLog);
+
+            _mockDbConnection = new Mock<IDbConnection>();
+
+            _certificateRepository = new CertificateRepository(_mockDbContext.Object,  _mockDbConnection.Object);
+        }
+        
+
+        [Test]
+        public async Task Then_Delete()
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs]: removed.

File: src/SFA.DAS.AssessorService.Application.Api/Controllers/CertificateController.cs [hunk 8]
```diff
@@ -107,5 +107,23 @@ public async Task<IActionResult> Approvals([FromBody] CertificateApprovalRequest
 
             return Ok();
         }
+
+        [HttpDelete("deletecertificate", Name = "DeleteCertificate")]
+        [SwaggerResponse((int)HttpStatusCode.OK, Type = typeof(Certificate))]
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Application.Api/Controllers/CertificateController.cs]: Removed the Type

File: src/SFA.DAS.AssessorService.Application.Api.Client/Clients/CertificateApiClient.cs [hunk 9]
```diff
@@ -84,5 +84,13 @@ public async Task<List<Option>> GetOptions(int stdCode)
                 return await RequestAndDeserialiseAsync<List<Option>>(httpRequest, "Could not get Options");
             }
         }
+
+        public async Task<Certificate> Delete(DeleteCertificateRequest deleteCertificateRequest)
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Application.Api.Client/Clients/CertificateApiClient.cs]: nope, Removed.

File: src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Data.UnitTests/Certificates/WhenSystemDeleteCertificate.cs:1]: @shomavg  above fyi.

**Pull request merged**